### DISCUSSION TITLE
Modifying album fields propagate to items?

### DISCUSSION THREAD

**Discussion by JOJ0** [beetbox/beets#4742] [Q&A] [answered]
Hi,
I'm wondering if I'm finding a bug or if this is a feature:

When I modify a bunch of albums to have a flexible attribute changed, it does not seem to change the flexible attributes of items. This inconsistency leads to a whole bunch of weird behaviour especially with path formats which I think respect both: item and album fields.

Ist this intendend? If yes, why does it make sense? IMHO it would be much cleaner if for example:

`beet mod -a toplevel=NewCategory toplevel:OldCategory`

would also change the items `toplevel` field.

Another example with a built-in flex attr:

`beet mod -a data_source=ManuallyTagged the unreleased album I did not find online`

The items' data_source field of this album now still are empty _or_ might be wrongly populated with `MusicBrainz` (or other tagging source) because of unsuccessful autotagging attempts. It would have been better if the "childs" (items) would now get the field changed as well.

Feature request? Bug report? Works as designed an should stay that way? Cheers :-)

Also an important fact is: On a `beet import --set=toplevel=NewCategory something new` BOTH the album and the items field are set! What I'm trying to say is that changing them afterwards is tedious.

**Comment by JOJ0** [comment]
Reminder to myself: Possibly related stuff. Read this later: 

https://github.com/beetbox/beets/pull/2988

https://github.com/beetbox/beets/pull/4663

**Comment by sampsyo** [comment]
Yes—for better or for worse, flexible attributes on albums are distinct from flexible attributes on items. I totally realize that this can lead to strange situations, but the reason it's not currently a _total_ disaster is that items get to "use" the flexible attributes from their albums in several contexts.

Namely, #2988 points at a situation where you use use `beet ls foo:x bar:y` and `$bar` (but not `$foo`) is only defined on the album. We now automatically "fall back" so items get to act as if they have the same metadata as their albums. This means that, most of the time, you can ignore the fact that the actual database associates the data differently with the two structures.

**Reply by JOJ0** [reply]
Thanks for clarifying and pointing out the fallback feature. That helps a lot with narrowing down what I actually don't like about it.

Yes, I came across the fallback feature recently and thought it's a good thing. And don't get me wrong, I do find the fact that item and album attrs are a different thing very useful. But often I find myself trying "to clean up the mess". I use flex attrs for broad categories of music and I just don't like the fact that when I change something for an album (mod -a) and in a next step I use a query for a specific track, I might get wrong information about that category because the item stayed untouched.

So in the end I often do almost the same command twice, just to make sure everything is clean, and that I find pointless. Eg.

```
beet mod -a cat=one
beet mod cat=one
```

I have several ideas how that could change:

- ~We have fallbacks. So on an `import` we would only set the album attribute but on an `import -s` we would set the item attribute.~ Update: After reading old issues, not an option. https://github.com/beetbox/beets/pull/3927#issuecomment-835153398
- Above idea but make it configurable. user can choose if regular album `import` sets album level or both album and item level.
- ~Change the behaviour of `beet mod -a`, so it always also changes the childs/items attributes.~ Update: The configurable option below is probably a better idea.
- The behaviour of `beet mod -a` is configurable. User can choose whether they want -a change items and albums, or just albums.

What do you think of them? Which one's you think could work? Which one's you think are problematic or breaking changes?

I think my personal favorites are 1 or 2.

**Reply by JOJ0** [reply]
After doin some "homework" and reading through discussions and issues I come to the conclusion that the only sensible way is option 2 or 4 which keep the default behaviour as-is.

With my "option 1" what I actually wanted is the "old" behaviour where imports only set album level fields (except on singelton imports). I still think that is the more logical soltuion, but users where asked for there opinion and decisions where made.

Funnily enough the "fallback" feature in #2988 (merged 7 March, 2021) came _before_ the `import --set` changes ablum+items in #3927 (merged 14 May 2021). I find that a little odd but anway...


With "option 3" I would probably run into the same problems. Some users by now are very used to this behaviour, a configurable solution as in "option 4" is more sensible.

I did not do my research yet for options 3 and 4......next step....let's see if there have been discussions around that already....

**Reply by sampsyo** [reply]
All this makes sense… but honestly at the moment, I can't 100% recall why "option 3" (i.e., permanently change `beet modify -a` to affect items too) is a bad idea. I totally believe this is one of those things where I previously thought about it harder and determined this was the case, but I don't immediately recall why it was the case. Anyway, if you have any great evidence for why that would be a terrible idea (other than that it would be surprising for existing users whose workflows depend on the ability to maintain separate album/item metadata), that would be awesome.

But yeah, barring that, I like the idea of making it configurable—so `beet modify -a` can *optionally* affect items too.

**Reply by JOJ0** [reply]
Thanks. Ok, if I get around to it, and I will, at some point, I'll look into submitting a PR. I would even go that far and change `-a` to affecting items too **as the default** but leave it configurable to get back the current behaviour (`-a` affecting album only). After reflecting your last message, I think that sounds like the most "correct" way and would let users that really require it, get back their usual workflows. Good?

**Reply by JOJ0** [reply]
Something I just noticed and want to share and document here. It seems `mod -a` _does_ also change items when the `added` field is modified (`a:yy-mm` in my album/item formats):

```
$ beet ls -a the prototype years
1997 - Grooverider - [487219 9] The Prototype Years  a:23-04/MusicBrainz b:844k [24/24]

$ beet ls the prototype years track:1
Grooverider - The Prototype Years - 01 - Dreams of Heaven feat. Codename John (1997/GB)  [487219 9] f:FLAC a:23-04/MusicBrainz |Drum And Bass, Jungle

$ beet mod -a added="2002-11-17 0:0:0" the prototype years 
Modifying 1 albums.
1997 - Grooverider - [487219 9] The Prototype Years  a:23-04/MusicBrainz b:844k [24/24]
  added: 2023-04-05 08:30:38 -> 2002-11-17 00:00:00
Really modify, move and write tags? (Yes/no/select) y

$ beet ls -a the prototype years                          
1997 - Grooverider - [487219 9] The Prototype Years  a:02-11/MusicBrainz b:844k [24/24]

$ beet ls the prototype years track:1                     
Grooverider - The Prototype Years - 01 - Dreams of Heaven feat. Codename John (1997/GB)  [487219 9] f:FLAC a:02-11/MusicBrainz |Drum And Bass, Jungle
```

In the database now:
```
album
1037487600.0

item
1037487600.0
```

**Reply by JOJ0** [reply]
I hope I don't confuse anything here: I have the importadded plugin enabled. I think it doesn't have to do with above behaviour but just mentioning it to make sure.

```
importadded:
    preserve_mtimes: yes
    preserve_write_mtimes: yes
```

**Reply by sampsyo** [reply]
I could be wrong, but I think this behavior (`beet modify -a` affects both albums and their items) is true for all fixed fields? For example, `beet modify albumartist=foo` would have the same behavior as modifying `added`.

**Reply by sampsyo** [reply]
(And I meant to say: yes! That plan sounds great!!!)

**Comment by JOJ0** [comment]
More stuff to read:

https://github.com/beetbox/beets/pull/3927

https://github.com/beetbox/beets/discussions/3925

https://github.com/beetbox/beets/issues/2800

**Comment by JOJ0** [comment, marked as answer]
Skimming thru old unanswered discussions... I think most of this is implemented in https://github.com/beetbox/beets/pull/4823
